### DISCUSSION TITLE
Spans and Traces by hand

### DISCUSSION THREAD

**Discussion by NameOfTheDragon** [datalust/seq-tickets#2219] [General]
Is it feasible to create spans and/or traces without having to use OpenTelemetry? I have a number of products that use NLog as the back end, but I do have an abstraction over the top of it. I was wondering if there's a way to add thsi capability to my NLog abstraction and create the spans and traces manually?

**Comment by KodrAus** [comment]
Hi @NameOfTheDragon :wave:

If you have control over the JSON payloads produced then you certainly can construct span events yourself. We've got some docs on the properties you need to include [here](https://docs.datalust.co/docs/posting-raw-events#reified-properties). Basically, you need to add the following properties to your events:

- `@st`: The start time of the span in the same format as `@t`
- `@tr`: The trace id as a 32 char hex string
- `@sp`: The span id as a 16 char hex string

You can also add extra properties in `@sc` and `@ra`, but these are optional from Seq's point of view.

As for manually constructing spans, you can do this using the `System.Diagnostics.Activity` APIs yourself. Even if you don't use the project directly, you might find some inspiration from how we do this in [SerilogTracing](https://github.com/serilog-tracing/serilog-tracing).

Hope that helps!
